File: ios/Mercury/MercuryKit/Notifications/NotificationVisibilityPolicy.swift
import Foundation
import MercuryCore

/// Facade over the shared KMP core's visibility policy. iOS has no separate
/// window-focus state, so windowFocused stays pinned to the shared default
/// (true) — semantics are unchanged from the previous two-field rule.
enum NotificationVisibilityPolicy {
    static func shouldPost(
        sessionID: String,
        visibility: SessionNotificationVisibility
    ) -> Bool {
        MercuryCore.NotificationVisibilityPolicy.shared.shouldPost(
            sessionId: sessionID,
            visibility: MercuryCore.SessionNotificationVisibility(
                appForeground: visibility.appForeground,
                windowFocused: true,
                visibleSessionId: visibility.visibleSessionID
            )
        )
    }
}

File: shared/mercury-core/src/commonMain/kotlin/com/unsupportedpastels/mercury/core/notifications/NotificationVisibilityPolicy.kt
package com.unsupportedpastels.mercury.core.notifications

/**
 * What the user can currently see, as policy input. [windowFocused] exists
 * for Android multi-window; iOS has no separate window-focus state and pins
 * it to the default `true`, which preserves its two-field semantics exactly.
 */
data class SessionNotificationVisibility(
    val appForeground: Boolean = false,
    val windowFocused: Boolean = true,
    val visibleSessionId: String? = null,
)

object NotificationVisibilityPolicy {
    /**
     * Post unless the user is actively looking at this session: app in the
     * foreground, window focused, and the same durable session on screen.
     */
    fun shouldPost(sessionId: String, visibility: SessionNotificationVisibility): Boolean = !(
        visibility.appForeground &&
            visibility.windowFocused &&
            visibility.visibleSessionId == sessionId
        )
}
